### DISCUSSION TITLE
Where to call "await AppInsights.SetAuthenticatedUserContext" on Blazor-Server with ASP.Net login

### DISCUSSION THREAD

**Discussion by wolfgangschneider** [IvanJosipovic/BlazorApplicationInsights#144] [Q&A] [answered]
HI 
pageViews works well except user_accountId.  I tryed out severals places but had only partial success. Is there a HowTo somewhere?
thanks

PS if it makes sence I can explayn exactly what I dis and the results

**Comment by IvanJosipovic** [comment, marked as answer]
Hey,

I had to do some more changes to make use ids work that I haven't documented yet.
The user id is set when the user authenticates, this is not done for subsequent loads of the app.

First, perform this https://github.com/IvanJosipovic/BlazorApplicationInsights#set-user-name
Then, add the following code to the App.razor.cs or MainLayout.razor.cs

```
using BlazorApplicationInsights;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Authorization;
using Microsoft.AspNetCore.Components.WebAssembly.Hosting;

namespace App.Client.Shared;

public partial class MainLayout
{
    [Inject] private ILogger<MainLayout> Logger { get; set; }

    [CascadingParameter] public Task<AuthenticationState> AuthenticationState { get; set; }

    [Inject] private IApplicationInsights AppInsights { get; set; }

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        var auth = await AuthenticationState;

        if (auth.User.Identity.IsAuthenticated)
        {
            var newName = auth.User.FindFirst("preferred_username")?.Value;

            Logger.LogInformation("{0} Logged In.", newName);

            // Replace illegal characters
            newName = new System.Text.RegularExpressions.Regex("[,;=| ]+").Replace(newName, "_");

            await AppInsights.SetAuthenticatedUserContext(newName, null, true);
        }
    }
}
```

Let me know if this helps.

**Comment by wolfgangschneider** [comment]
Thanks for quick reply.
It works fine  on first look on MainLayout (not on App). I will test more scenario's over weekend and let you know.

Thanks Wolfgang
